weave/integrations/langchain/langchain.py

```diff
@@ -367,6 +371,131 @@ class WeaveTracer:  # type: ignore
         pass
 
 
+def _on_finish_handler(
+    call: Call, output: Any, exception: Optional[BaseException] = None
+) -> None:
+    """Handle the finish event of a call.
+
+    Args:
+        call: The call that finished
+        output: The output of the call
+    """
+    usage: Union[dict[str, dict[str, int]], None] = None
+    if output is not None and "outputs" in output and len(output["outputs"]) > 0:
+        first_output = output["outputs"][0]
+        if (
+            (message := safe_get(first_output, ["output"]))
+            and hasattr(message, "response_metadata")
+            and safe_get(message.response_metadata, ["token_usage"])
+        ):
+            # LangChain AIMessage response format
+            usage = _extract_chat_message_usage(message.response_metadata)
+        elif (generations := safe_get(first_output, ["generations"])) and (
+            len(generations) > 0
+            and len(generations[0]) > 0
+            and safe_get(
+                generations[0][0],
+                ["message", "kwargs", "response_metadata", "token_usage"],
+            )
+        ):
+            # openai response format
+            usage = _extract_openai_usage(output)
+        elif (generations := safe_get(first_output, ["generations"])) and (
+            len(generations) > 0
+            and len(generations[0]) > 0
+            and safe_get(
+                generations[0][0],
+                ["message", "kwargs", "response_metadata", "usage_metadata"],
+            )
+        ):
+            # google vertexai response format
+            usage = _extract_google_vertexai_usage(output)
+        elif (generations := safe_get(first_output, ["generations"])) and (
+            len(generations) > 0
+            and len(generations[0]) > 0
+            and safe_get(generations[0][0], ["generation_info", "usage_metadata"])
+        ):
+            # google genai response format
+            usage = _extract_google_genai_usage(output)
+
+    if call.summary is not None and usage is not None:
+        call.summary.update({"usage": usage})
+
+
+def _extract_openai_usage(output: dict) -> dict:
```

Comment:
i'd recommend typing these extractions as `LLMUsageSchema` as defined in `trace_server_schema` to ensure they conform to the expected payload